plugins/wacom-i2c/fu-wac-emr-device.c

```diff
@@ -0,0 +1,279 @@
+		return FALSE;
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_wac_emr_device_erase_codemem (FuWacEmrDevice *self,
+				 guint8 idx,
+				 guint8 block_nr,
+				 GError **error)
+{
+	guint8 rsp[RSP_SIZE];
+	guint8 cmd[] = {
+		FU_WAC_DEVICE_BL_REPORT_ID_CMD,
+		FU_WAC_DEVICE_BL_CMD_ERASE_FLASH,
+		idx,			/* echo */
+		block_nr,		/* erased block */
+		0xff,			/* checksum */
+	};
+	cmd[4] = fu_wac_emr_device_calc_checksum (0x05, 0x07, cmd, 4);
+	if (!fu_wac_device_cmd (FU_WAC_DEVICE (self),
+				cmd, sizeof(cmd), rsp, sizeof(rsp), 50,
+				FU_WAC_DEVICE_CMD_FLAG_POLL_ON_WAITING, error)) {
+		g_prefix_error (error, "failed to erase codemem: ");
+		return FALSE;
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_wac_emr_device_write_block (FuWacEmrDevice *self,
+			       guint32 idx,
+			       guint32 address,
+			       const guint8 *data,
+			       guint16 datasz,
+			       GError **error)
+{
+	guint blocksz = fu_wac_device_get_block_sz (FU_WAC_DEVICE (self));
+	g_autofree guint8 *cmd = g_malloc0 (blocksz + 8 + 2);
+	cmd[0] = FU_WAC_DEVICE_BL_REPORT_ID_CMD;
+	cmd[1] = FU_WAC_DEVICE_BL_CMD_WRITE_FLASH;
+	cmd[2] = (guint8) idx;	/* echo */
+
+	/* address */
+	fu_common_write_uint32 (cmd + 3, address, G_LITTLE_ENDIAN);
+
+	/* size */
+	if (datasz != blocksz) {
+		g_set_error (error,
+			     G_IO_ERROR,
+			     G_IO_ERROR_FAILED,
+			     "block size 0x%x != 0x%x untested",
+			     datasz, (guint) blocksz);
+		return FALSE;
+	}
+	cmd[7] = datasz / 8;
+
+	/* data */
+	memcpy (cmd + 8, data, datasz);
+
+	/* cmd and data checksums */
+	cmd[8 + blocksz + 0] = fu_wac_emr_device_calc_checksum (0x05, 0x4c, cmd, 8);
+	cmd[8 + blocksz + 1] = fu_wac_emr_device_calc_checksum (0x00, 0x00, data, datasz);
+
+	/* set, no get */
+	if (!fu_wac_device_set_feature (FU_WAC_DEVICE (self),
+					FU_WAC_DEVICE_BL_REPORT_ID_CMD,
+					cmd, sizeof(cmd), error)) {
+		g_prefix_error (error, "failed to write codemem: ");
+		return FALSE;
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_wac_emr_device_write_firmware (FuDevice *device, GPtrArray *chunks, GError **error)
+{
+	FuWacEmrDevice *self = FU_WAC_EMR_DEVICE (device);
+	guint8 idx = 0;
+
+	/* check MPU type: FIXME, move? */
+	if (!fu_wac_emr_device_check_mpu_type (self, error))
+		return FALSE;
+
+	/* erase */
+	fu_device_set_status (device, FWUPD_STATUS_DEVICE_ERASE);
+	if (!fu_wac_emr_device_erase_datamem (self, error))
+		return FALSE;
+	for (guint i = 127; i >= 8; i--) {
+		if (!fu_wac_emr_device_erase_codemem (self, idx++, i, error))
+			return FALSE;
+	}
+
+	/* write */
+	fu_device_set_status (device, FWUPD_STATUS_DEVICE_WRITE);
+	for (guint i = 0; i < chunks->len; i++) {
+		FuChunk *chk = g_ptr_array_index (chunks, i);
+		if (fu_wac_common_block_is_empty (chk->data, chk->data_sz))
+			continue;
+		if (!fu_wac_emr_device_write_block (self,
+						    chk->idx,
+						    chk->address,
+						    chk->data,
+						    chk->data_sz,
+						    error))
+			return FALSE;
+		g_usleep (50);
```

Comment:
Does this really need to settle between every write?  Or is it just after the 3 writes before the read?

Reply:
I just copied the reference code from Wacom, I don't like the`"WWWRRRWWWRRR` pattern at all, so I've asked Wacom if I can just use `fu_wac_device_cmd()` like all the other commands.

Reply:
I've massively simplified this now I have docs...